### PR TITLE
Initial Code Migration

### PR DESCRIPTION
**Pull request by wbaldoumas**
## Description

This set of changes migrates the internal library code, examples, tests, and docs into the open source repository.

## Type of Change

- [ ] Bug Fix
- [ ] New Feature
- [ ] Breaking Change
- [ ] Refactor
- [ ] Documentation
- [x] Other (please describe) - migration of internal to external repository

## Checklist

- [x] I have read the [contributing guidelines](https://github.com/wayfair-incubator/AeroSharp/blob/main/CONTRIBUTING.md)
- [x] Existing issues have been referenced (where applicable)
- [x] I have verified this change is not present in other open pull requests
- [x] Functionality is documented
- [x] All code style checks pass
- [x] New code contribution is covered by automated tests
- [x] All new and existing tests pass

### CODE REVIEW

File: docs/README.md [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+
+IEnumerable<MyDataType> listResult = await list.ReadAllAsync(CancellationToken.None);
+// listResult contains [ MyDataType("list item 1"), MyDataType("list item 2") ]
+```
+
+## Client Provider
+
+Aerospike provides a [client](https://www.aerospike.com/docs/client/csharp/) that maintains a pool of connections to an Aerospike cluster. The client needs to know the location of at least one node in the cluster and can discover the remaining nodes. The client can also accept username/password credentials as well as other parameters (e.g. connection timeout).
+
+This library uses a [client provider](./src/AeroSharp/Connection/IClientProvider.cs) to give the data access objects access to an Aerospike client. Use the `ClientProviderBuilder` to configure and build a client provider, e.g.:
```
Comment (jdhmtl) [inline docs/README.md:65]:
```suggestion
This library uses a [client provider](../src/AeroSharp/Connection/IClientProvider.cs) to give the data access objects access to an Aerospike client. Use the `ClientProviderBuilder` to configure and build a client provider, e.g.:
```

File: docs/README.md [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+    .List.RemoveByValue("list_1_bin", listItem)
+    .List.Append("list_2_bin", listItem)
+    .SizeAsync("list_2_bin", cancellationToken); // Return the result of the "size" operation
+```
+
+### Batch Operations
+
+_This feature is not yet implemented._
+
+Aerospike does not support batching update or write operations on multiple records in the same request. It does, however, support [fetching batches of records](https://docs.aerospike.com/docs/guide/batch.html#known-limitations). The [KeyValueStore](#keyvaluestore) data access object fetches records in batches when reading multiple keys of the same type.
```
Comment (jdhmtl) [inline docs/README.md:421]: Should this be removed from the documentation until it is implemented?

File: docs/README.md [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+await keyValueStore.ReadModifyWriteAsync(
+    "some_key", 
+    addOperation, 
+    updateOperation, 
+    timeToLive: TimeSpan.FromHours(5), 
+    CancellationToken.None
+);
+```
+
+This mimics the AddOrUpdate method on the ConcurrentDictionary object in C#. In this way, the user does not have to care whether the record already exists or not: the library will handle creating or updating the record appropriately.
```
Comment (jdhmtl) [inline docs/README.md:339]:
```suggestion
This mimics the [AddOrUpdate method on the ConcurrentDictionary](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.addorupdate?view=net-6.0#System_Collections_Concurrent_ConcurrentDictionary_2_AddOrUpdate__0__1_System_Func__0__1__1__) object in C#. In this way, the user does not have to care whether the record already exists or not: the library will handle creating or updating the record appropriately.
```

File: examples/AeroSharp.Examples.HelloWorld/Program.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+using System.Threading.Tasks;
+using AeroSharp.Connection;
+using AeroSharp.DataAccess;
+
+namespace AeroSharp.Examples.HelloWorld
+{
+    /// <summary>
+    /// The simplest example of connecting to a local instance of Aerospike and writing/reading some data.
+    /// 
+    /// Before running this, run `start_aerospike_in_docker.sh` to start a preconfigured local Aerospike instance.
```
Comment (jdhmtl) [inline examples/AeroSharp.Examples.HelloWorld/Program.cs:11]: That shell script currently lives in the integration tests directory. Given that it's also in use here, should we consider moving it to a `scripts` or `utilities` directory? Either way, the reference should probably be updated so people know where to find said script.

File: src/AeroSharp/DataAccess/ReadConfiguration.cs [hunk 5]
```diff
@@ -0,0 +1,86 @@
+﻿using System;
+using System.Diagnostics.CodeAnalysis;
+
+namespace AeroSharp.DataAccess
+{
+    /// <summary>
+    /// /// <summary>
+    /// A class for storing and passing the configurable settings for reads.
+    /// </summary>
+    /// </summary>
```
Comment (jdhmtl) [inline src/AeroSharp/DataAccess/ReadConfiguration.cs:10]: That doesn't look quite right

File: tests/AeroSharp.Tests/Utility/TestPreparer.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        private const int Port = 3000;
+
+        public const string TestSet = "test_set";
+        public const string TestNamespace = "test_namespace";
+        public static DataAccess.DataContext TestDataContext => new(TestNamespace, TestSet);
+
+        /// <summary>
+        /// Call this method in the Setup phase of unit tests to clear out test data and get a ClientProvider and DataAccessConfiguration.
+        /// </summary>
+        /// <returns>Tuple containing a client provider and data access configuration.</returns>
```
Comment (jdhmtl) [inline tests/AeroSharp.Tests/Utility/TestPreparer.cs:45]: Let's ensure this is documented in any testing documentation we have / create.

**Review by jdhmtl — COMMENTED** [review]
Looks good! Mostly small comments.

File: docs/README.md (hunk 2 above)
Comment (wbaldoumas) [reply docs/README.md:421]: I think it should be, yeah. Nice catch!

File: src/AeroSharp/DataAccess/ReadConfiguration.cs (hunk 5 above)
Comment (wbaldoumas) [reply src/AeroSharp/DataAccess/ReadConfiguration.cs:10]: Hah, fun. I'll clean this up.

**Review by jdhmtl — APPROVED** [review]

**Pull request merged**